---
title: "Low-quality Cells Validation"
output: html_document
---

```{r global_options, include=FALSE}
library(ggplot2)
library(ggsci)
library(ggpubr)
library(dplyr)
library(parallel)
library(reshape2)
library(pagoda2)
library(ComplexHeatmap)
library(circlize)
library(RColorBrewer)

source("./Functions/PlotFuncs.R")
source("./Functions/Functions.R")

knitr::opts_chunk$set(fig.width=5, fig.height=3, echo=FALSE, warning=FALSE, message=FALSE)

theme_set(theme_base)

# kDatasetName <- 'SCG71'
kDatasetName <- 'pbmc33k'
kPlotsFolder <- paste0('/d0-mendel/home/viktor_petukhov/Data/Plots/PaperReview/LowQualityCells/', kDatasetName, '/')
```

```{r}
GetPagoda <- function(cm) {
  r <- Pagoda2$new(cm, modelType='plain', trim=5, n.cores=10)
  r$adjustVariance(plot=F, do.par=F, gam.k=10)
  
  r$calculatePcaReduction(nPcs=100,n.odgenes=1000, maxit=1000)
  r$makeKnnGraph(k=30,type='PCA', center=T,distance='cosine',weight.type='none');
  r$getKnnClusters(method=infomap.community,type='PCA',name='infomap')
  
  r$getEmbedding(type='PCA', perplexity=30, embeddingType = 'tSNE')
  return(r)
}
```

```{r}
# holder <- readRDS('/d0-mendel/home/viktor_petukhov/Data/SCG71/est_11_14_poisson_real/SCG71.rds')
holder <- readRDS('/d0-mendel/home/viktor_petukhov/Data/10x/pbmc33k/est_11_14/pbmc33k.rds')
```

```{r}
PlotCellsNumberLine(holder$aligned_umis_per_cell, estimate.cells.number=T) + theme_pdf()
```

```{r}
kRealCellsNum <- EstimateCellsNumber(holder$aligned_umis_per_cell)$expected
```

```{r}
# scores_mit_filt <- ScorePipelineCells(holder, filter.high.mit.fraction=T, mit.chromosome.name='MT')
# scores <- ScorePipelineCells(holder, filter.high.mit.fraction=F, mit.chromosome.name='MT', predict.all=T)
scores <- ScorePipelineCells(holder, predict.all=T)
mit_chromosome_fraction <- GetChromosomeFraction(holder$reads_per_chr_per_cells$Exon[colnames(holder$cm_raw),], 'chrM')
```

```{r}
r_cm <- holder$cm_raw[, order(Matrix::colSums(holder$cm_raw), decreasing=T)]
cell_mask <- setNames(rep(FALSE, ncol(r_cm)), colnames(r_cm))
cell_mask[names(scores)[scores > 0.6]] <- TRUE
cell_mask[1:kRealCellsNum] <- TRUE

r_cm <- r_cm[grep("^[^;]+$", rownames(r_cm)), cell_mask]
r_cm <- r_cm[, Matrix::colSums(r_cm) > 50]
r_cm <- r_cm[Matrix::rowSums(r_cm) > 10, ]
r_cm <- r_cm[, Matrix::colSums(r_cm) > 0]
dim(r_cm)
```

```{r}
r <- GetPagoda(r_cm)
```

```{r, fig.width=4, fig.height=3}
gp_clusters <- PlotPagodaEmbeding(r, clustering.type='infomap', show.legend=F, min.cluster.size=70, mark.clusters=T, alpha=0.5, size=0.5, font.size=4.5) + 
  theme_pdf(F)
gp_clusters
```

```{r, fig.width=4, fig.height=3}
gp_umis_per_cb <- PlotPagodaEmbeding(r, colors=Matrix::colSums(r_cm), show.legend=T, alpha=0.5, size=0.5) + 
  theme_pdf(F, legend.pos=c(0, 0)) +
  guides(color=guide_colorbar(title='#Molecules per cell', direction='horizontal', barwidth=unit(1.5, 'in'), title.position='top', title.hjust=0, raster=T)) +
  scale_color_distiller(trans='log', breaks=c(100, 1000, 5000), palette='YlGn') +
  theme(legend.margin = ggplot2::margin(t=0, r=0, b=0, l=0.05, unit='in'))
gp_umis_per_cb
```

```{r, fig.width=4, fig.height=3}
gp_mit <- PlotPagodaEmbeding(r, colors=mit_chromosome_fraction, show.legend=T, alpha=0.5, size=0.5, show.ticks=F) + 
  scale_color_distiller(values=c(0, 0.05, 0.1, 0.3, 1.0), palette='RdYlBu') + 
  theme_pdf(show.ticks=F, legend.pos=c(0, 0)) + 
  guides(color=guide_colorbar(title='Mitochondrial fraction', direction='horizontal', barwidth=unit(1.5, 'in'), title.position='top', raster=T)) +
  theme(legend.margin = ggplot2::margin(t=0, r=0, b=0, l=0.05, unit='in'))
gp_mit
```

```{r, fig.width=4, fig.height=3}
is_real <- as.integer(scores > 0.6)
names(is_real) <- names(scores)

gp_score <- PlotPagodaEmbeding(r, colors=scores, show.legend=T, min.cluster.size=100, alpha=0.5, size=0.5) + 
  scale_color_distiller(values=c(1.0, 0.8, 0.7, 0.6, 0.0), palette='RdYlBu', guide=guide_colourbar(title='Score', raster=T), breaks=seq(0.1, 0.9, length.out=3)) + 
  theme_pdf(F, c(0, 0)) +
  guides(color=guide_colorbar(title='Score', direction='horizontal', barwidth=unit(1, 'in'), title.position='top', title.hjust=0.5)) +
  theme(legend.margin = ggplot2::margin(t=0, r=0.05, b=0, l=0.05, unit='in'))
gp_score
```

```{r, fig.width=8, fig.height=5}
cowplot::plot_grid(plotlist=lapply(list(gp_clusters, gp_umis_per_cb, gp_mit, gp_score), 
                                   function(x) x + rremove("xylab") + theme(plot.margin=margin(l=0.01, r=0.01, b=0.01, t=0.01, "in"))), 
                   labels = c("A", "B", "C", "D"), nrow=2, ncol=2, label_x=0.01, label_y=0.99)

ggsave(filename=paste0(kPlotsFolder, 'fig_clusters.pdf'), width=8, height=5, units='in')
```

```{r}
umis_per_cb <- Matrix::colSums(holder$cm_raw)
is_real_simple <- as.integer(umis_per_cb >= sort(umis_per_cb, decreasing=T)[kRealCellsNum])
names(is_real_simple) <- names(umis_per_cb)
# is_real_simple <- is_real_simple[names(is_real)]
# PlotPagodaEmbeding(r, clusters=factor(is_real_simple), show.legend=T, min.cluster.size=100, alpha=0.5, size=0.5) +
#   scale_color_npg() + theme_pdf(F, c(1, 1))
```

```{r}
real_clusters_num <- 17
printTableCell <- function(is.included) {
  return(paste0(round(mean(is.included == 0) * 100, 1), '%', ' (', sum(is.included == 0), ')'))
}

getErrorsData <- function(is.real, clusters, trash.clusts) {
  res <- split(is.real, clusters[names(is.real)])
  res$X <- as.integer(unlist(res[trash.clusts])) - 1
  return(res)
}

real_clusts <- c(paste0(1:real_clusters_num), 'X')
trash_clusts <- paste0((real_clusters_num + 1):max(as.integer(r$clusters$PCA$infomap)))

errors_data <- list(Simple=getErrorsData(is_real_simple, r$clusters$PCA$infomap, trash_clusts),
                    KDE=getErrorsData(is_real, r$clusters$PCA$infomap, trash_clusts))

cluster_reduction <- as.data.frame(lapply(errors_data, function(l) sapply(l[real_clusts], printTableCell)))
rownames(cluster_reduction) <- real_clusts

write.csv(cluster_reduction, file=paste0(kPlotsFolder, 'ExcludedFraction.csv'), row.names=T)
cluster_reduction
```

# Heatmaps
```{r}
is_rescued <- is_real & !is_real_simple[names(is_real)]
is_filtered <- !is_real & is_real_simple[names(is_real)]

cell_type <- as.factor(sapply(paste0(is_rescued - is_filtered), 
                              switch, `1`="Rescued", `0`="Base", `-1`="Filtered"))
names(cell_type) <- names(is_rescued)
```

```{r, message=FALSE, warning=FALSE}
selected_clusters <- c(1, 2, 5, 13) #c(4, 5, 8)
tested_clusts <- sort(r$clusters$PCA$infomap[r$clusters$PCA$infomap %in% selected_clusters])
tested_clusts <- setNames(as.integer(tested_clusts), names(tested_clusts))
tested_clusts <- tested_clusts[intersect(names(tested_clusts), names(is_real))]
tested_clusts <- tested_clusts[is_real[names(tested_clusts)] | is_real_simple[names(tested_clusts)]]

umis_per_cb_subset <- log10(Matrix::colSums(r_cm[, names(tested_clusts)]))
tested_clusts <- tested_clusts[order(tested_clusts, cell_type[names(tested_clusts)], -umis_per_cb_subset)]

r$getDifferentialGenes(groups=tested_clusts, type='PCA', clusterType = 'infomap')
de_genes <- Reduce(union, lapply(r$diffgenes$PCA$customClustering, function(x) rownames(x)[1:min(100, nrow(x))]))
de_genes <- de_genes[!is.na(de_genes)]

m_subset <- log10(1e-6 + as.matrix(r$counts[names(tested_clusts), de_genes]))
annotation_df <- data.frame(row.names=names(tested_clusts), 
                            CellType=cell_type[names(tested_clusts)],
                            UmisPerCell=umis_per_cb_subset[names(tested_clusts)],
                            Cluster=tested_clusts)

h_clust_col <- hclust(dist(t(m_subset)))
```

```{r}
PrepareBaseHeatmap <- function(mtx, annotation.df, h.clust.columns, show_heatmap_legend=T, row_title='', column_title='', column_dend=F) {
  hm.base <- Heatmap(mtx[rownames(annotation.df),], 
                   cluster_rows=F, cluster_columns=h.clust.columns, name = "log10(expression)", 
                   show_heatmap_legend=show_heatmap_legend, show_row_names = F, show_column_names = F, show_column_dend=column_dend,
                   column_dend_height=unit(0.3, 'in'),
                   heatmap_legend_param=list(legend_direction = "horizontal", legend_width=unit(2, 'in')),
                   col=colorRampPalette(rev(brewer.pal(n = 7, name = "RdYlBu")))(100),
                   row_title=row_title, column_title=column_title, column_title_side='bottom')
  return(hm.base)
}

PrepareHeatmapAnnotation <- function(annotation.df, colors, show_legend=F) {
  annotation <- HeatmapAnnotation(df=annotation.df, which = "row", 
                                 annotation_legend_param = list(
                                   UmisPerCell = list(title = "log10(#molecules per cell)", legend_direction = "horizontal", 
                                                      color_bar='continuous', legend_width=unit(2, 'in')), 
                                   Cluster = list(title = "Cluster", legend_direction = "horizontal", nrow=1, legend_gap=unit(c(1, 1), 'in'))),
                                 col=colors, show_legend=show_legend)
  
  return(annotation)
}
```

```{r, fig.width=8, fig.height=4}
annotation_df_s <- split(annotation_df, annotation_df$CellType)

colors_lst <- list(UmisPerCell=colorRamp2(breaks=seq(min(annotation_df$UmisPerCell), max((annotation_df$UmisPerCell)), length.out=100), 
                                          colors=colorRampPalette(brewer.pal(n = 9, name = "OrRd"))(100)),
                   Cluster=setNames(colorRampPalette(rev(brewer.pal(n = length(selected_clusters), name = "Dark2")))(length(selected_clusters)), unique(annotation_df$Cluster)))

annotation_df_cur <- annotation_df_s$Base %>% dplyr::select(-CellType)
hm_base <- PrepareBaseHeatmap(m_subset, annotation_df_cur, h_clust_col, show_heatmap_legend=F, column_dend=T, row_title='Cells', column_title='')
ha_base <- PrepareHeatmapAnnotation(annotation_df_cur, colors_lst, show_legend=F)

annotation_df_cur <- annotation_df_s$Filtered %>% dplyr::select(-CellType)
hm_filt <- PrepareBaseHeatmap(m_subset, annotation_df_cur, h_clust_col, show_heatmap_legend=T, column_dend=F, row_title='Cells', column_title='Genes')
ha_filt <- PrepareHeatmapAnnotation(annotation_df_cur, colors_lst, show_legend=T)

annotation_df_cur <- annotation_df_s$Rescued %>% dplyr::select(-CellType)
hm_rescued <- PrepareBaseHeatmap(m_subset, annotation_df_cur, h_clust_col, show_heatmap_legend=F, column_dend=T, row_title='', column_title='Genes')
ha_rescued <- PrepareHeatmapAnnotation(annotation_df_cur, colors_lst, show_legend=F)
```

```{r, fig.width=8, fig.height=8}
# pdf(paste0(kPlotsFolder, 'filtered_genes_heatmap.pdf'), width=8, height=8, paper='special')
# png(paste0(kPlotsFolder, 'filtered_genes_heatmap.png'), width=8, height=8, units='in', res=300)
draw(hm_base + ha_base, padding = unit(c(4.02, 0, 0, 3.9), "in"))
draw(hm_rescued + ha_rescued, newpage=F, padding = unit(c(3.66, 4.2, 0, 0), "in"))
draw(hm_filt + ha_filt, newpage=F, padding = unit(c(0, 0, 4.02, 1.83), "in"), heatmap_legend_side='right', annotation_legend_side='right')
# dev.off()
```

# Pagoda biomarkers
```{r}
# gene_names <- read.table('/d0-mendel/home/viktor_petukhov/Data/10x/pbmc33k/filtered_gene_bc_matrices/hg19/genes.tsv', row.names=1)
# gene_names <- setNames(gene_names$V2, rownames(gene_names))
```

```{r}
r$getDifferentialGenes(type='PCA',clusterType = 'infomap', upregulated.only=T, z.threshold=4)
sapply(r$diffgenes$PCA$infomap[1:20], nrow)
markers <- as.data.frame(lapply(r$diffgenes$PCA$infomap[1:20], function(df) as.vector(na.omit(gene_names[rownames(df)]))[1:50]))
names(markers)[sapply(markers, function(m) 'IL7R' %in% m)]
```

# Pagoda application
```{r}
# Generate GO genesets for the web app
go_sets <- p2.generate.human.go.web(colnames(r$counts))

# Generate differental expression between each cluster and everything else
# Load these clusters as pre-defined gene sets with the given prefix
de_sets <- get.de.geneset(r, groups = r$clusters$PCA$infomap, prefix = 'de_')

# Merge Genesets
go_sets <- c(go_sets, de_sets)

# Additional metadata generation
additional_metadata <- list()
additional_metadata$altCluster <- p2.metadata.from.factor(r$clusters$PCA$infomap, displayname = 'Infomap', s = 0.7, v = 0.8, start = 0, end = 0.5)

# Generate and display web app
r_web_object <-
  make.p2.app(
    r,
    dendrogramCellGroups = r$clusters$PCA$infomap,
    additionalMetadata = additional_metadata,
    geneSets = go_sets,
    show.clusters = FALSE, # Hide the clusters that were used for the dendrogram from the metadata
  )
```

```{r}
show.app(app=r_web_object, name=paste0('p2', kDatasetName))
```
