---
title: "10x Aml035 Post"
output:
  html_document:
    df_print: kable
    number_sections: yes
    theme: cerulean
    toc: yes
  html_notebook: default
---

```{r global_options, include=FALSE}
library(ggplot2)
library(ggsci)
library(ggpubr)
library(ggrastr)
library(dplyr)
library(parallel)
library(reshape2)

source("./Functions/PlotFuncs.R")
source("./Functions/Functions.R")

knitr::opts_chunk$set(fig.width=5, fig.height=3, echo=FALSE, warning=FALSE, message=FALSE)

theme_set(theme_base)
kPlotsFolder <- '~/Data/Plots/Paper/Review/UmiErrors/'
kDataPath <- '/d0-mendel/home/viktor_petukhov/Data/10x/aml035_post_transplant/est_10_20_umi_quality/'
```

```{r}
# holder <- readRDS(paste0(kDataPath, '/aml035ost_transplant.rds'))
# reads_per_umi_per_cell <- holder$reads_per_umi_per_cell
# reads_per_umi_per_cell$reads_per_umi <- FilterNUmis(holder$reads_per_umi_per_cell$reads_per_umi)
#
# saveRDS(reads_per_umi_per_cell, paste0(kDataPath, 'reads_per_umi_per_cell.rds'))

reads_per_umi_per_cell <- readRDS(paste0(kDataPath, 'reads_per_umi_per_cell.rds'))
# raw <- readRDS(paste0(kDataPath, 'raw.rds'))

length(reads_per_umi_per_cell$reads_per_umi)
```

# UMI errors
```{r}
tr_cur <- TrimAndCorrect(reads_per_umi_per_cell, umi.trim.length=6, mc.cores.large=30, mc.cores.small=30, verbosity.level=1, prepare.only=T)
classifier <- TrainNBClassifier(tr_cur$reads.per.umi.per.cb, distribution.smooth=50, umi.probabilities=tr_cur$umi.probabilities, mc.cores=10)
```

```{r}
umi_probabilities_map <- new(CppMap, names(tr_cur$umi.probabilities), as.vector(tr_cur$umi.probabilities))
# gp <- profvis::profvis(r0 <- lapply(tr_cur$reads.per.umi.per.cb[tr_cur$umis.per.gene > 1][1:1000], FilterUmisInGene, umi_probabilities_map, classifier,
#        tr_cur$correction.info$dp.matrices, tr_cur$correction.info$neighb.prob.index, tr_cur$collisions.info))

r0 <- lapply(tr_cur$reads.per.umi.per.cb[tr_cur$umis.per.gene > 1][1:1000], FilterUmisInGene, umi_probabilities_map, classifier,
       tr_cur$correction.info$dp.matrices, tr_cur$correction.info$neighb.prob.index, tr_cur$collisions.info)
```

```{r}
r0 <- mclapply(tr_cur$reads.per.umi.per.cb[1:100000], FilterUmisInGene, umi_probabilities_map, classifier,
               tr_cur$correction.info$dp.matrices, tr_cur$correction.info$neighb.prob.index,
               tr_cur$collisions.info, mc.cores=30)
```

```{r}
largest_gene <- tr_cur$reads.per.umi.per.cb[[which.max(tr_cur$umis.per.gene)]]
r1 <- FilterUmisInGene(largest_gene, umi_probabilities_map, classifier, tr_cur$correction.info$dp.matrices, 
                       tr_cur$correction.info$neighb.prob.index, tr_cur$collisions.info, verbose=T)
```

```{r}
trimmed <- lapply(6:9, function(i) TrimAndCorrect(reads_per_umi_per_cell, umi.trim.length=i, mc.cores.large=30, mc.cores.small=30, verbosity.level=1))
for (i in 1:length(trimmed)) {
  trimmed[[i]]$umis.per.gene.adj <- trimmed[[i]]$collisions.info[trimmed[[i]]$umis.per.gene]
}

raw <- list()

raw$umi_distribution <- GetUmisDistribution(reads_per_umi_per_cell$reads_per_umi, smooth = 10)
raw$umi_probabilities <- raw$umi_distribution / sum(raw$umi_distribution)

max_umi_per_gene <- sapply(reads_per_umi_per_cell$reads_per_umi, length) %>% max()
raw$collisions_info <- FillCollisionsAdjustmentInfo(raw$umi_probabilities, max_umi_per_gene)
raw$correction_info <- PrepareUmiCorrectionInfo(umi.probabilities=raw$umi_probabilities, raw$collisions_info[max_umi_per_gene])

raw$filt_umis_per_gene_simple <- CorrectUmiSequenceErrors(reads_per_umi_per_cell, raw$umi_probabilities, method='Classic',
                                                          correction.info=raw$correction_info, collisions.info=raw$collisions_info,
                                                          mc.cores=30, verbosity.level=2, return='umis')

raw$umis_per_gene <- sapply(reads_per_umi_per_cell$reads_per_umi, length)

saveRDS(raw, paste0(kDataPath, 'raw4_umi.rds'))
saveRDS(trimmed, paste0(kDataPath, 'trimmed4_umi.rds'))
```

```{r}
raw <- readRDS(paste0(kDataPath, 'raw3.rds'))
trimmed <- readRDS(paste0(kDataPath, 'trimmed3.rds'))
```

## Plots
## Supp. figure
```{r}
plots <- lapply(6:9, function(i) PlotTrimmedCorrections(trimmed[[i - 5]], raw, i, log=T, rast.width=4,
                                                        rast.height=4.5, rast.dpi=200, heights.ratio=c(2.5, 3)))
large_plots <- lapply(plots, `[[`, 'large')
large_plots <- lapply(large_plots, `+`, annotation_logticks(sides='b'))
small_plots <- lapply(plots, `[[`, 'small')
```

```{r, fig.width=8, fig.height=9, warning=FALSE, message=FALSE}
gp_large <- BuildPanel4(large_plots, "Corrected #UMIs without trimming", "Error on trimmed data, %", show.legend=T)
gp_small <- BuildPanel4(small_plots, "Corrected #UMIs without trimming", "Error on trimmed data, %", show.legend=F)
supp_fig_errors <- cowplot::plot_grid(gp_large, gp_small, nrow=2, labels=c('A', 'B'), rel_heights=c(5, 6))
supp_fig_errors
```

```{r}
pdf(paste0(kPlotsFolder, 'SuppFigTrimRast.pdf'), width=8, height=9)
print(supp_fig_errors)
dev.off()
```

## Main figure
```{r}
umis_per_gene <- as.data.frame(lapply(trimmed, `[[`, 'umis.per.gene'))
colnames(umis_per_gene) <- paste0(6:9)
errors <- cbind((umis_per_gene - raw$filt_umis_per_gene_simple) / raw$filt_umis_per_gene_simple, Real=raw$filt_umis_per_gene_simple)
errors <- melt(errors, id.vars='Real', variable.name='UmiLength', value.name='Error')
errors_sum <- errors %>% group_by(Real, UmiLength) %>% summarise(Error=mean(Error, trim=0.2))
```

```{r}
kPlotWidth <- 8
kPlotHeight <- 7
gp1 <- ggplot(errors_sum, aes(x=Real, y=Error, color=UmiLength)) +
  geom_point_rast(size=0.3, alpha=0.5, width=kPlotWidth / 2, height=kPlotHeight / 3) +
  geom_smooth() +
  scale_x_log10(expand=c(0.01, 0)) + annotation_logticks(sides='b') + ylim(-0.8, 0.75) +
  theme_pdf(legend.pos=c(0, 0)) +
  scale_color_npg() +
  guides(color=guide_legend(title='Length of trimmed UMI', ncol=4))
gp1
```

```{r}
fig_umi_length <- 6

raw$filt_reads_per_umi_simple <- CorrectUmiSequenceErrors(reads_per_umi_per_cell, raw$umi_probabilities, method='Classic',
                                                          correction.info=raw$correction_info, collisions.info=raw$collisions_info,
                                                          mc.cores=30, verbosity.level=2, return='reads')

trimmed_corrected_cells <- lapply(raw$filt_reads_per_umi_simple, TrimUmis, fig_umi_length)
```

```{r}
umis_per_gene_plt <- data.frame(none=sapply(trimmed_corrected_cells, length), real=raw$filt_umis_per_gene_simple)

umis_per_gene_plt$empirical <- trimmed[[fig_umi_length - 5]]$collisions.info[umis_per_gene_plt$none]
umis_per_gene_plt$uniform <- sapply(umis_per_gene_plt$none, AdjustGeneExpressionClassic, 4^fig_umi_length)
umis_per_gene_plt_df <- melt(as.data.frame(umis_per_gene_plt), id.vars='real', variable.name='Adjustment', value.name='value')

umis_per_gene_plt_df <- umis_per_gene_plt_df %>% group_by(Adjustment, real) %>% summarize(value=mean(value, 0.2))

gp2 <- ggplot(umis_per_gene_plt_df, aes(x=real, y=100 * (value - real) / real, col=Adjustment)) +
  geom_point_rast(size=0.3, alpha=0.3, width=kPlotWidth / 2, height=kPlotHeight / 3) +
  geom_smooth() +
  scale_x_log10(expand=c(0.01, 0)) + annotation_logticks(sides='b') +
  scale_y_continuous(expand=c(0, 0), limits=c(-61, 11)) +
  scale_color_nejm() +
  labs(x='Corrected #molecules without trimming', y='Error, %') +
  guides(color = guide_legend(title='Collisions adjustment', ncol=3)) +
  theme_pdf(legend.pos=c(0, 0))

gp2
```

```{r}
correction_colors <- c("#2222F2", "#017A5A", "#9B3BB8", "#E69F00", "#BD5500", '#757575')
color_scale <- scale_color_manual(name='Correction', values=alpha(correction_colors, 0.7),
                                  labels=c('Cell Ranger', 'Bayesian', 'cluster', 'cluster-neq', 'directional', 'no correction'))
```

```{r}
fig_umi_length <- 7
corrected_data <- as.data.frame(trimmed[[fig_umi_length - 5]]$filt_cells)
corrected_data$NoCorrection <- trimmed[[fig_umi_length - 5]]$umis.per.gene
corrected_data$real <- raw$filt_umis_per_gene_simple

plot_df <- melt(corrected_data, id.vars='real', variable.name='Correction') %>%
  group_by(Correction, real) %>% summarize(value=median(value))

gp3 <- ggplot(plot_df, aes(x=real, y=100 * (value - real) / real, color=Correction)) +
  geom_point_rast(size=0.3, alpha=0.3, width=kPlotWidth / 2, height=kPlotHeight / 3) +
  geom_smooth() +
  scale_x_log10(expand=c(0.01, 0)) + annotation_logticks(sides='b') +
  scale_y_continuous(limits=c(-101, 51), expand=c(0, 0)) +
  scale_color_manual(values=correction_colors[2:length(correction_colors)]) +
  guides(color=guide_legend(ncol=3)) + theme_pdf(legend.pos=c(0, 0))

gp3
```

```{r}
aml035_plots <- list(gp1=gp1, gp2=gp2, gp3=gp3)
saveRDS(aml035_plots, paste0(kDataPath, '/gg_plots.rds'))
```

```{r}
f_bmmc_plots <- readRDS('/d0-mendel/home/viktor_petukhov/Data/10x/frozen_bmmc_healthy_donor1/est_07_05/correlations_by_clust.rds')
f_pbmc_plots <- readRDS('/d0-mendel/home/viktor_petukhov/Data/10x/frozen_pbmc_donor_a/est_07_10/correlations_by_clust.rds')
```

```{r, fig.width=5, fig.height=4}
f_bmmc_plots$gg <- ggplot(f_bmmc_plots$correlations_by_clust, aes(x=spearman, color=CorrectionType, linetype=CollisionsAdjustment)) +
  geom_density(size=1.3, show.legend=F) +
  geom_line(aes(y=rep(0, nrow(f_bmmc_plots$correlations_by_clust))), size=1.3) +
  color_scale +
  scale_linetype_manual(name='Collisions\nadjustment', values=c('solid', 'dashed'), labels=c('adjusted', 'unadjusted')) +
  scale_x_continuous(expand=c(0, 0), limits=c(0.55, 0.85)) +
  scale_y_continuous(expand=c(0, 0), limits=c(0, 9.5), name='Density') +
  theme_pdf(legend.pos=c(0, 1)) +
  guides(linetype=guide_legend(order=2, keywidth=1.2, override.aes=list(linetype=c('solid', '21'))),
         color=guide_legend(reverse=T, order=1, keywidth=1.2))
f_bmmc_plots$gg
```

```{r, fig.width=5, fig.height=8}
corr_fig <- cowplot::plot_grid(f_bmmc_plots$gg + rremove("xlab"),
                      f_pbmc_plots$gg + scale_x_continuous(expand=c(0, 0), limits=c(0.75, 0.9)) +
                        color_scale + rremove("xlab") + rremove("legend"),
                      ncol=1, nrow=2, labels=c('D', 'E'), align='v')

corr_fig_ann <- annotate_figure(corr_fig + theme(plot.margin=margin(t=10, unit='pt')),
                                bottom=text_grob("Spearman correlation", size=14))
corr_fig_ann
```

```{r, fig.width=4, fig.height=7}
scale_color_short <- scale_color_manual(values=alpha(correction_colors[2:length(correction_colors)], 0.7))
figure_trim <- cowplot::plot_grid(aml035_plots$gp1 + theme_pdf() + ylab('Error, %') + rremove("xlab"),
                         aml035_plots$gp2 + theme_pdf() + rremove("xlab"),
                         aml035_plots$gp3 + theme_pdf() + ylab('Error, %') +
                           scale_color_short +
                           guides(color=guide_legend(title='Correction (see Figure D)', label.theme=element_blank(), ncol=5, reverse=T)) +
                           rremove("xlab"),
                         ncol=1, nrow=3, labels = c("A", "B", "C"), align='v')

figure_trim_ann <- annotate_figure(figure_trim + theme(plot.margin=margin(t=10, l=0, unit='pt')),
                                   bottom=text_grob("Gene expression magnitude (#molecules)", size=14))

figure_trim_ann
```

```{r, fig.width=8, fig.height=7}
full_fig <- cowplot::plot_grid(figure_trim_ann + theme(plot.margin=margin()), corr_fig_ann + theme(plot.margin=margin()), ncol=2)
full_fig_ann <- annotate_figure(full_fig, fig.lab="Figure 2", fig.lab.pos="top.right")
# ggexport(full_fig_ann, filename=paste0(kPlotsFolder, '/Figure.pdf'), pdf.width=8)
full_fig_ann
```

```{r}
pdf(paste0(kPlotsFolder, '/Figure.pdf'), width=8, height=7)
print(full_fig_ann)
dev.off()
```
