---
title: "10x Aml035 Post"
output:
  html_document:
    df_print: kable
    number_sections: yes
    theme: cerulean
    toc: yes
  html_notebook: default
---

```{r global_options, include=FALSE}
library(ggplot2)
library(ggsci)
library(ggpubr)
library(ggrastr)
library(dplyr)
library(parallel)
library(reshape2)

source("./Functions/PlotFuncs.R")
source("./Functions/Functions.R")
source("./Functions/NewClassification.R")

knitr::opts_chunk$set(fig.width=5, fig.height=3, echo=FALSE, warning=FALSE, message=FALSE)

theme_set(theme_base)
kPlotsFolder <- '~/Data/Plots/Paper/Review/UmiErrors/'
kDataPath <- '/d0-mendel/home/viktor_petukhov/Data/10x/aml035_post_transplant/est_10_20_umi_quality/'
```

```{r}
# holder <- readRDS(paste0(kDataPath, '/aml035ost_transplant.rds'))
# reads_per_umi_per_cell <- holder$reads_per_umi_per_cell
# reads_per_umi_per_cell$reads_per_umi <- FilterNUmis(holder$reads_per_umi_per_cell$reads_per_umi)
# 
# saveRDS(reads_per_umi_per_cell, paste0(kDataPath, 'reads_per_umi_per_cell.rds'))

# reads_per_umi_per_cell <- readRDS(paste0(kDataPath, 'reads_per_umi_per_cell.rds'))
# length(reads_per_umi_per_cell$reads_per_umi)
```
```{r}
# reads_per_umi_per_cell$reads_per_umi <- reads_per_umi_per_cell$reads_per_umi[1:100000]
```

```{r}
raw <- readRDS(paste0(kDataPath, 'raw.rds'))
trimmed <- readRDS(paste0(kDataPath, 'trimmed_6.rds'))
```

```{r}
raw_corrected_rpus <- CorrectUmiSequenceErrorsClassic(raw$correction_info, mult=1, mc.cores=30)
```

# Reads per UMI
```{r}
reads_per_umi_train <- ReadsPerUmiDataset(trimmed$reads.per.umi.per.cb, max.umis.per.cb=4, mc.cores=10)
sum_rpus <- colSums(reads_per_umi_train)

(read_error_prob <- sum_rpus['Small'] / sum(sum_rpus))
(theta <- EstimateReadsBetaBinomial(reads_per_umi_train$Small, reads_per_umi_train$Large, read_error_prob))
```

```{r}
reads_num <- 10
ggplot() + 
  geom_histogram(aes(x=reads_per_umi_train %>% filter(Large + Small == reads_num) %>% .$Small, y=..count.. / sum(..count..), color='Observed')) +
  geom_bar(aes(x=0:10, y=emdbook::dbetabinom(0:10, size=reads_num, prob=read_error_prob, theta=theta), color='beta'), stat='identity', width=0.5, alpha=0.5)
```

```{r}
rpus_extracted <- ExtractReadsPerUmi(trimmed$reads.per.umi.per.cb, mc.cores=10)
reads_num <- (max(rpus_extracted %>% sapply(max)) * 1.5) %>% round()
umi_num <- nchar(names(trimmed$umi.probabilities)[1]) * 3
probs_given_reads_large <- ErrorProbsGivenNumOfReadsLarge(reads_num, read_error_prob, umi_num)

rpu_distribution <- SmoothDistribution(unlist(rpus_extracted) - 1, 10, log.probs=F)
```

```{r}
classifier <- TrainNBClassifier(trimmed$reads.per.umi.per.cb, 50, trimmed$correction.info, trimmed$umi.probabilities)
classifier$RpU <- list(Distribution=log(rpu_distribution), NegDistParams=list(Theta=theta, Prob=read_error_prob), ErrNumRL=probs_given_reads_large)
classifier$MaxAdjacentUmisNum <- names(trimmed$umi.probabilities)[1] %>% nchar() * 3
```

```{r}
classifier_df <- data.frame(Base=names(trimmed$umi.probabilities)[1:10], Target='AAAAAA', Quality = 52, MinRpU=1, MaxRpU=20)
PredictNew(classifier, classifier_df)
```

```{r}
umi_probabilities_map <- new(CppMap, names(trimmed$umi.probabilities), as.vector(trimmed$umi.probabilities))
```

```{r}
rpus_filt <- lapply(trimmed$reads.per.umi.per.cb[1:1000], FilterUmisInGeneNew, umi_probabilities_map, 
                    classifier, trimmed$correction.info$neighbours.per.umi, trimmed$correction.info$dp.matrices, 
                    trimmed$correction.info$neighb.prob.index, trimmed$collisions.info)

rpus_filt <- mclapply(trimmed$reads.per.umi.per.cb, FilterUmisInGeneNew, umi_probabilities_map, 
                      classifier, trimmed$correction.info$neighbours.per.umi, trimmed$correction.info$dp.matrices, 
                      trimmed$correction.info$neighb.prob.index, trimmed$collisions.info, mc.cores=30)

trimmed$filt_cells$NB <- trimmed$collisions.info[sapply(rpus_filt, length)]
```

```{r}
umi_probabilities_map <- new(CppMap, names(trimmed$umi.probabilities), as.vector(trimmed$umi.probabilities))
# gene0 <- trimmed$reads.per.umi.per.cb[trimmed$umis.per.gene == 50][[2]]
gene0 <- trimmed$reads.per.umi.per.cb[trimmed$umis.per.gene >= 2000][[2]]
r0 <- FilterUmisInGeneNew(gene0, umi_probabilities_map, classifier, trimmed$correction.info$neighbours.per.umi, 
                          trimmed$correction.info$dp.matrices, trimmed$correction.info$neighb.prob.index, trimmed$collisions.info)
```


# Validation plots
```{r}
ggs <- PlotTrimmedCorrections(trimmed, raw, 6, raster=F)
ggs$large + annotation_logticks(sides='b') + theme_pdf()
ggs$small + theme_pdf()
```

```{r}
ggs <- PlotTrimmedCorrections(trimmed, raw, 6, raster=F)
ggs$large + annotation_logticks(sides='b') + theme_pdf()
ggs$small + theme_pdf()
```

```{r}
ggs <- PlotTrimmedCorrections(trimmed, raw, 6, raster=F)
ggs$large + annotation_logticks(sides='b') + theme_pdf()
ggs$small + theme_pdf()
```

```{r}
ggs <- PlotTrimmedCorrections(trimmed, raw, 6, raster=F)
ggs$large + annotation_logticks(sides='b') + theme_pdf()
ggs$small + theme_pdf()
```

# Comparison with answers
```{r}
length(raw_corrected_rpus[trimmed$umis.per.gene >= 2000][[2]])
real_umis <- names(raw_corrected_rpus[trimmed$umis.per.gene >= 2000][[2]]) %>% substr(start=1, stop=6) %>% unique()
c(length(gene0), length(real_umis))
```

```{r}
tryCatch(umi_probabilities <- umi_probabilities_map$at(names(gene0)), error=function(x) stop(paste0("Error: ", x$message)))

normalizers <- lapply(names(gene0), GetAdjacentUmis)%>% lapply(umi_probabilities_map$at) %>%
  sapply(sum) %>% setNames(names(gene0))

classifier_df <- PredictNew(classifier, PrepareClassifierData(gene0, umi_probabilities, normalizers), gene0, 
                            trimmed$correction.info$neighbours.per.umi, trimmed$correction.info$dp.matrices, 
                            trimmed$correction.info$neighb.prob.index, trimmed$collisions.info[length(gene0)])

classifier_df$IsErroneous <- !(classifier_df$Base %in% real_umis)
```

```{r}
c(sum(classifier_df$IsErroneous & classifier_df$IsMerged), 
  sum(classifier_df$IsErroneous & !classifier_df$IsMerged),
  sum(!classifier_df$IsErroneous & classifier_df$IsMerged))
```

```{r}
dfs <- setNames(split(classifier_df, classifier_df$IsErroneous), c('Real', 'Error'))
ggplot(mapping=aes(y=..count.. / sum(..count..))) +
  geom_histogram(aes(dfs$Real$MaxRpU, color='Real'), bins=100, alpha=0.6) + 
  geom_histogram(aes(dfs$Error$MaxRpU, color='Error'), bins=100, alpha=0.6) + 
  # geom_histogram(aes(dfs$Error$MaxRpU + dfs$Error$MinRpU, color='Sum'), bins=100, alpha=0.4) + 
  geom_histogram(aes(reads_per_umi_train$Large, color='Total'), bins=100, alpha=0.4) + 
  theme_pdf(legend.pos=c(1, 1))
```

```{r}
ggplot(mapping=aes(y=..count.. / sum(..count..))) +
  geom_histogram(aes(dfs$Real$MinRpU, fill='Real'), bins=100, alpha=0.6) + 
  geom_histogram(aes(dfs$Error$MinRpU, fill='Error'), bins=100, alpha=0.6) + 
  theme_pdf(legend.pos=c(1, 1))
```

```{r}
ggplot(mapping=aes(y=..count.. / sum(..count..))) +
  geom_histogram(aes(dfs$Real$Quality, fill='Real'), bins=100, alpha=0.6) + 
  geom_histogram(aes(dfs$Error$Quality, fill='Error'), bins=100, alpha=0.6) + 
  theme_pdf(legend.pos=c(1, 1))
```

```{r}
ggplot(mapping=aes(y=..count.. / sum(..count..))) +
  geom_histogram(aes(dfs$Real$Nucleotides, fill='Real'), bins=100, alpha=0.6) + 
  geom_histogram(aes(dfs$Error$Nucleotides, fill='Error'), bins=100, alpha=0.6) + 
  theme_pdf(legend.pos=c(1, 1))

ggplot(mapping=aes(y=..count.. / sum(..count..))) +
  geom_histogram(aes(dfs$Real$Position, fill='Real'), bins=100, alpha=0.6) + 
  geom_histogram(aes(dfs$Error$Position, fill='Error'), bins=100, alpha=0.6) + 
  theme_pdf(legend.pos=c(1, 1))
```

```{r}
extractAdjacentUmisNum <- function(x) names(x) %>% substr(1, 6) %>% SubsetAdjacentUmis() %>% sapply(length)
not_single_raw_rpus <- raw_corrected_rpus[sapply(raw_corrected_rpus, length) > 1]

adj_umi_n_per_gene <- mclapply(not_single_raw_rpus, extractAdjacentUmisNum, mc.cores=10)
```

```{r}
adj_df <- mclapply(adj_umi_n_per_gene, function(g) tibble(ProbId=trimmed$correction.info$neighb.prob.index[names(g)], AdjNum=g, GeneSize=length(g)), mc.cores=30)
adj_df <- adj_df %>% bind_rows()
```

```{r}
# (plot_gs <- min(adj_df$GeneSize[adj_df$GeneSize > 200]))
plot_gs <- 30
(s_n <- adj_df %>% filter(GeneSize == plot_gs) %>% .$ProbId %>% table())
(plot_pi <- names(which.max(s_n)))
```

```{r}
max_adj_num <- min(18, plot_gs)
ggplot() +
  geom_histogram(data=(adj_df %>% filter(GeneSize == plot_gs, ProbId == plot_pi)), 
                 mapping=aes(x=AdjNum, y = ..count.. / sum(..count..), fill='Observed'), bins=50) +
  geom_bar(aes(x=0:max_adj_num, y=trimmed$correction.info$dp.matrices[[plot_pi]][1:(max_adj_num + 1), trimmed$collisions.info[plot_gs]], fill='Estimated'), 
           stat='identity', width=0.3, alpha=0.6) +
  theme_pdf(legend.pos=c(1, 1))
```

## Sampling
```{r}
# sampled_pairs <- mclapply(1:100000, function(i) sample(names(trimmed$umi.probabilities), size=2, replace=T, prob=trimmed$umi.probabilities), mc.cores=10)
sampled_pairs <- mclapply(1:50000, function(i) sample(names(raw$umi_probabilities), size=2, replace=T, prob=raw$umi_probabilities), mc.cores=30)
sampled_pairs <- lapply(sampled_pairs, substr, 1, 6)
  
sampled_adj_umi_n <- mclapply(sampled_pairs, function(x) SubsetAdjacentUmis(x) %>% sapply(length), mc.cores=5)
sampled_adj_df <- mclapply(sampled_adj_umi_n, function(g) tibble(ProbId=trimmed$correction.info$neighb.prob.index[names(g)], AdjNum=g, GeneSize=length(g)), mc.cores=30) %>% bind_rows()
sampled_adj_df <- sampled_adj_df %>% filter(GeneSize == 2)

trimmed$correction.info$dp.matrices[[plot_pi]][1:2, 2]
table(sampled_adj_df$AdjNum) / nrow(sampled_adj_df)
```

```{r}
max_adj_num <- 2
ggplot() +
  geom_histogram(data=(adj_df %>% filter(GeneSize == 2, ProbId == plot_pi)), 
                 mapping=aes(x=AdjNum, y = ..count.. / sum(..count..), fill='Observed'), bins=50) +
  geom_histogram(data=(sampled_adj_df %>% filter(ProbId == plot_pi)), 
                 mapping=aes(x=AdjNum, y = ..count.. / sum(..count..), fill='Sampling'), bins=50, alpha=0.7) +
  geom_bar(aes(x=0:max_adj_num, y=trimmed$correction.info$dp.matrices[[plot_pi]][1:(max_adj_num + 1), trimmed$collisions.info[2]], fill='Estimated'), 
           stat='identity', width=0.3, alpha=0.6) +
  theme_pdf(legend.pos=c(1, 1))
```

## Raw data analysis
```{r}
raw_inds <- which(sapply(raw_corrected_rpus, length) > 1)

pair_mask <- sapply(adj_umi_n_per_gene, length) == 2
adj_mask <- adj_umi_n_per_gene[pair_mask] %>% sapply(max) == 1

raw_inds <- raw_inds[pair_mask][adj_mask]
sum(adj_mask) / sum(pair_mask)
```

```{r}
# paired_raw_rpus <- ExtractReadsPerUmi(raw_corrected_rpus[raw_inds])
# paired_raw_rpus <- paired_raw_rpus[sapply(paired_raw_rpus, length) == 2]

paired_raw_rpus <- ExtractReadsPerUmi(raw_corrected_rpus[sapply(raw_corrected_rpus, length) == 2])

edit_distances <- lapply(paired_raw_rpus, names) %>% lapply(sapply, strsplit, '') %>% sapply(function(x) sum(x[[1]] != x[[2]]))
diff_positions <- lapply(paired_raw_rpus, names) %>% lapply(sapply, strsplit, '') %>% lapply(function(x) which(x[[1]] != x[[2]]))

table(unlist(diff_positions[edit_distances > 1 & edit_distances < 4])) %>% (function(x) x / sum(x))
```

```{r}
table(edit_distances) / length(edit_distances)
```

