---
title: "10x Aml035 Post"
output:
  html_document:
    df_print: kable
    number_sections: yes
    theme: cerulean
    toc: yes
  html_notebook: default
---

```{r}
source("./Functions/PlotFuncs.R")

LogSumExp<- function(x) {
  # Computes log(sum(exp(x))
  # Uses offset trick to avoid numeric overflow: http://jblevins.org/notes/log-sum-exp
  if ( max(abs(x)) > max(x) )
    offset <- min(x)
  else
    offset <- max(x)
  return(log(sum(exp(x - offset))) + offset)
}

GetProbabilityPartsDf <- function(clf, classifier.df, gene.size, return.diff=F) {
  nucl.prob.err <- log(clf$Negative$Nucleotides[classifier.df$Nucleotides + 1])
  position.prob.err <- log(clf$Negative$Position[classifier.df$Position + 1])

  min.rpu.prob.err <- emdbook::dbetabinom(classifier.df$MinRpU - 1, size=classifier.df$MinRpU + classifier.df$MaxRpU - 1,
                                          prob=clf$Negative$MinRpU$prob, theta=clf$Negative$MinRpU$theta, log=T)
  # max.rpu.prob.err <- dnbinom(classifier.df$MaxRpU - 1, size=clf$Negative$MaxRpU$size, mu=clf$Negative$MaxRpU$mu, log=T)

  quantized.quality <- Quantize(classifier.df$Quality, clf$QualityQuantBorders)

  min.rpu.prob <- log(clf$Common$RpuProbs[classifier.df$MinRpU])
  # max.rpu.prob <- log(clf$Common$RpuProbs[classifier.df$MaxRpU])

  quality.prob <- log(clf$Common$Quality[quantized.quality + 1]);
  quality.prob.err <- log(clf$Negative$Quality[quantized.quality + 1]);

  umi.prob <- log(1 - (1 - classifier.df$UmiProb)^gene.size)
  
  if (return.diff) {
    res <- list(nucl.prob.err + position.prob.err - umi.prob, min.rpu.prob.err - min.rpu.prob, 
                quality.prob.err - quality.prob) # max.rpu.prob.err - max.rpu.prob, 
    res_names <- c('umi', 'min.rpu', 'quality') #, 'max.rpu'
  } else {
    res <- list(nucl.prob.err, position.prob.err, umi.prob, min.rpu.prob.err, min.rpu.prob, quality.prob.err, #, max.rpu.prob.err
              quality.prob) #, max.rpu.prob
    res_names <- c('nucl.prob.err', 'position.prob.err', 'umi.prob', 'min.rpu.prob.err', 'min.rpu.prob', 'quality.prob.err', 
                   'quality.prob')
  }

  res <- data.frame(lapply(res, as.vector))
  colnames(res) <- res_names
  return(bind_cols(classifier.df[,1:2], res))
}

FilterNUmis <- function(reads.per.umi) {
  return(lapply(reads.per.umi, function(rpus) rpus[grep("^[^N]+$", names(rpus))]))
}

GetExprDf <- function(filt.umis.per.gene.vec, real.umis.per.gene.vec, raw.umis.per.gene.vec, adjustment=FALSE) {
  expressions <- as.data.frame(filt.umis.per.gene.vec)
  expressions$Adjustment <- adjustment
  expressions$RealValue <- real.umis.per.gene.vec
  expressions$NoCorrection <- raw.umis.per.gene.vec
  return(expressions %>% filter(NoCorrection > 1))
}

GetPlotExrDf <- function(trimmed.cur, raw, umi.length, return.all=FALSE) {
  expr.df <- GetExprDf(trimmed.cur$filt_cells, raw$filt_umis_per_gene_simple, trimmed.cur$umis.per.gene, adjustment=T) %>% 
    reshape2::melt(id.vars=c('RealValue', 'Adjustment'), variable.name='Correction') %>% mutate(UmiLen=umi.length)
  
  if (return.all)
    return(expr.df)
  
  return(expr.df %>% 
    group_by(Correction, RealValue, Adjustment) %>% 
    summarise(Max=quantile(value, 0.95), Min=quantile(value, 0.05), Median=median(value)))
}
```

```{r global_options, include=FALSE}
library(ggplot2)
library(ggsci)
library(ggpubr)
library(dplyr)
library(parallel)

knitr::opts_chunk$set(fig.width=5, fig.height=3, echo=FALSE, warning=FALSE, message=FALSE)

theme_set(theme_bw(base_size=14, base_family='Helvetica') + theme(plot.title = element_text(hjust = 0.5)))
```

```{r}
holder <- readRDS('/d0-mendel/home/viktor_petukhov/Data/10x/aml035_post_transplant/est_10_20_umi_quality/aml035ost_transplant.rds')
reads_per_umi_per_cell <- holder$reads_per_umi_per_cell
reads_per_umi_per_cell$reads_per_umi <- FilterNUmis(holder$reads_per_umi_per_cell$reads_per_umi)
# reads_per_umi_per_cell$reads_per_umi <- holder$reads_per_umi_per_cell$reads_per_umi[1:100000]
# saveRDS(reads_per_umi_per_cell, '/d0-mendel/home/viktor_petukhov/Data/10x/aml035_post_transplant/est_10_20_umi_quality/reads_per_umi_subset.rds')
```

```{r}
reads_per_umi_per_cell <- readRDS('/d0-mendel/home/viktor_petukhov/Data/10x/aml035_post_transplant/est_10_20_umi_quality/reads_per_umi_subset.rds')
# reads_per_umi_per_cell$reads_per_umi <- reads_per_umi_per_cell$reads_per_umi[1:1000]
```

<!-- Simple filtration of the raw data -->
```{r}
raw <- list()

raw$umi_distribution <- GetUmisDistribution(reads_per_umi_per_cell$reads_per_umi, smooth = 10)
raw$correction_info <- PrepareUmiCorrectionInfoWrapper(reads_per_umi_per_cell$reads_per_umi, umi.probabilities=raw$umi_distribution / sum(raw$umi_distribution),
                                                       method='Classic', verbosity.level=2, max.umi.per.gene=NULL, probability.quants.num=NULL)

raw$filt_umis_per_gene_simple <- CorrectUmiSequenceErrors(reads_per_umi_per_cell, raw$umi_distribution, method='Classic',
                                                          correction.info=raw$correction_info,
                                                          mc.cores=20, verbosity.level=2, return='umis')

trimmed <- TrimAndCorrect(reads_per_umi_per_cell, umi.trim.length=6, collisions.adj.step=30, mc.cores.large=20, mc.cores.small=20, verbosity.level=2, prepare.only=F)
```

```{r}
trimmed_reads_per_umi_per_cb <- reads_per_umi_per_cell
trimmed_reads_per_umi_per_cb$reads_per_umi <- trimmed$reads.per.umi.per.cb

trimmed$filt_cells$NB_p05 <- CorrectUmiSequenceErrors(trimmed_reads_per_umi_per_cb, umi.probabilities=trimmed$umi.probabilities,
                                                   collisions.info=trimmed$collisions.info, correction.info=trimmed$correction.info,
                                                   mc.cores=30, return='umis', verbosity.level=2, error.prior.prob=0.05)
```

## Plots
```{r}
plot_df <- GetPlotExrDf(trimmed, raw, 6) %>% filter(Correction != "NoCorrection" & Adjustment | Correction == "NoCorrection" & !Adjustment) %>% dplyr::select(-Adjustment)
plot_df_all <- GetPlotExrDf(trimmed, raw, 6, T) %>% filter(Correction != "NoCorrection" & Adjustment | Correction == "NoCorrection" & !Adjustment)
```

```{r, fig.width=9, fig.height=5}
p0 <- plot_df_all %>% filter(RealValue < 50)
ggplot(p0, aes(x=as.factor(ceiling(RealValue / 5) * 5) , y=100 * (value - RealValue) / RealValue, color=Correction)) + 
  geom_boxplot_jitter_outlier(outlier.jitter.width=NULL, outlier.jitter.height=1, outlier.size=0.7, outlier.alpha=0.7) +
  labs(x='Corrected #UMIs without trimming', y='Error on trimmed data, %') + 
  scale_linetype_manual(values=c('dashed', 'solid')) + scale_color_jco() +
  ylim(-50, 50) +
  theme(legend.position=c(1, 0), legend.justification=c(1, 0), strip.background =element_blank(), strip.text = element_text(size=14)) + 
  guides(color=guide_legend(ncol=3))
```

```{r, fig.width=7.5, fig.height=5}
ggplot(plot_df, aes(x=RealValue, y=100 * (Median - RealValue) / RealValue, color=Correction)) + 
  geom_point(size=0.3, alpha=0.3) + geom_smooth(alpha=0.1) + 
  labs(x='Corrected #UMIs without trimming', y='Error on trimmed data, %') + 
  scale_linetype_manual(values=c('dashed', 'solid')) + scale_color_jco() +
  scale_x_continuous(expand=c(0, 0), limits=c(1, 7999), trans='identity') + scale_y_continuous(expand=c(0, 0), limits=c(-101, 70)) + 
  theme(legend.position=c(0, 0), legend.justification=c(0, 0), strip.background =element_blank(), strip.text = element_text(size=14)) + 
  guides(color=guide_legend(ncol=3))
```
