---
title: "Filtration of low-quality cells"
author: "Petukhov Viktor"
date: "`r Sys.Date()`"
output: 
  rmarkdown::html_vignette:
    css:
vignette: >
  %\VignetteIndexEntry{Filtration of low-quality cells}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include=FALSE}
ggplot2::theme_set(ggplot2::theme_bw(base_size = 14) + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5)))
```

## Quick start
```{r}
library(dropestr)
data('lq_cells_data')
pipeline_data <- lq_cells_data$pipeline
mit_genes <- lq_cells_data$genes
```


The simplest way to score the cells for data, obtained with the pipeline:
```{r, fig.width=5, fig.height=4}
scores <- ScorePipelineCells(pipeline_data)
PlotCellScores(scores, y.threshold=0.9)
```

Cells with high mitochondrial fraction are probably dead, so it's reasonable to filter them out. There are two ways of 
distinguishing of mitochondrial reads: by chromosome name and by genesets. The first approach estimates the fraction of reads,
while the second works with UMIs. However, results are quite similar.

```{r, fig.width=4.5, fig.height=3}
scores_chromosome_filt <- ScorePipelineCells(pipeline_data, filter.high.mit.fraction=T, mit.chromosome.name='chrM')
scores_geneset_filt <- ScorePipelineCells(pipeline_data, filter.high.mit.fraction=T, mitochondrion.genes=mit_genes)
PlotCellScores(scores_chromosome_filt, y.threshold=0.9, main='Chromosome')
PlotCellScores(scores_geneset_filt, y.threshold=0.9, main='Geneset')
```

Answers are the same for `r round(mean((scores_chromosome_filt > 0.9) == (scores_geneset_filt > 0.9)) * 100, 2)`% cells.

## Manual filtration
This filtration can be done manually in more flexible way.
The first step is feature extraction from existed data.
```{r}
lq_cells_df <- PrepareLqCellsDataPipeline(pipeline_data, mit.chromosome.name="chrM")
```

Next, we need to estimate approximate number of real cells. It can be done using one of the following plots,
each of which shows the expected number of cells, however for different datasets some of them can give more precise result than the other:
```{r, fig.width=5, fig.height=3}
PlotCellsNumberLogLog(pipeline_data$aligned_umis_per_cell, estimate.cells.number=T)
PlotCellsNumberLine(pipeline_data$aligned_umis_per_cell, estimate.cells.number=T)
PlotCellsNumberHist(pipeline_data$aligned_umis_per_cell, estimate.cells.number=T)
```

Some cells can die prior to the sequencing, and they should be filtered. Such cell can be distinuished by mitochondrial fraction.
As was mentioned above, this fraction can be estimated using the chromosome name or the list of genes.
```{r, fig.width=4, fig.height=3}
mit_chromosome_fraction <- GetChromosomeFraction(pipeline_data$reads_per_chr_per_cells$Exon[colnames(pipeline_data$cm),], 'chrM')
mit_geneset_fraction <- GetGenesetFraction(pipeline_data$cm, mit_genes)

FractionSmoothScatter(mit_chromosome_fraction, plot.threshold=T, main='Chromosome')
FractionSmoothScatter(mit_geneset_fraction, plot.threshold=T, main='Geneset')
```

The last preparation step is obtaining initial labeling from the data. Here we can vary both borders of high/low quality cells and 
the level of mitochondrial fraction, which is considered to be a marker of dead cells.
```{r}
cells_number_manual <- list(min=300, max=800)

cells_quality <- EstimateCellsQuality(pipeline_data$aligned_umis_per_cell)
cells_quality_manual <- EstimateCellsQuality(pipeline_data$aligned_umis_per_cell, cells.number=cells_number_manual)

cells_quality_filt <- FilterMitochondrionCells(mit_geneset_fraction, cells_quality)
cells_quality_filt_manual <- FilterMitochondrionCells(mit_geneset_fraction, cells_quality, mit.threshold=0.2)
```

Scoring of the cells for different types of initial labeling:
```{r}
scores <- ScoreCells(lq_cells_df, cells_quality)
scores_manual <- ScoreCells(lq_cells_df, cells_quality_manual)

scores_filt <- ScoreCells(lq_cells_df, cells_quality_filt)
scores_filt_manual <- ScoreCells(lq_cells_df, cells_quality_filt_manual)
```

It can be seen that choice of threshold for mitochondrial fraction influence the result, but borders selection doesn't really affect set of cells, which were classified as high-quality.
```{r, fig.width=5, fig.height=4}
cells_number <- EstimateCellsNumber(pipeline_data$aligned_umis_per_cell)
PlotCellScores(scores, cells_number, 0.95, main='Default parameters')
PlotCellScores(scores_manual,cells_number_manual, 0.95, main='Manual borders')

PlotCellScores(scores_filt, cells_number, 0.95, main='Mit. filtration')
PlotCellScores(scores_filt_manual, cells_number, 0.95, main='Manual mit. filtration')
```


