# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

GetCrossmergedMask <- function(base_umis, target_umis) {
    .Call('_dropestr_GetCrossmergedMask', PACKAGE = 'dropestr', base_umis, target_umis)
}

SubsetAdjacentUmis <- function(umis) {
    .Call('_dropestr_SubsetAdjacentUmis', PACKAGE = 'dropestr', umis)
}

#' Fill information about adjacent UMIs, their probabilities and differences for each UMI.
#'
#' @param umi_probabilites vector of UMI probabilities.
#' @param adjacent_only logical, return only the list of adjacent UMIs.
#' @param show_progress show progress bar.
#' @return List with the information about adjacent UMIs.
#'
#' @export
FillAdjacentUmisData <- function(umi_probabilites, adjacent_only = FALSE, show_progress = FALSE) {
    .Call('_dropestr_FillAdjacentUmisData', PACKAGE = 'dropestr', umi_probabilites, adjacent_only, show_progress)
}

GetAdjacentUmisNum <- function(reads_per_umi_from, reads_per_umi_to, neighbourhood, total = TRUE, larger = FALSE, smaller = FALSE) {
    .Call('_dropestr_GetAdjacentUmisNum', PACKAGE = 'dropestr', reads_per_umi_from, reads_per_umi_to, neighbourhood, total, larger, smaller)
}

#' @export
FillDpMatrix <- function(prior_prob, neighbours_num, max_umi_per_cell) {
    .Call('_dropestr_FillDpMatrix', PACKAGE = 'dropestr', prior_prob, neighbours_num, max_umi_per_cell)
}

GetSmallerNeighboursDistributionsBySizes <- function(dp_matrices, larger_neighbours_num, neighbour_prob_inds, size_adj, max_neighbour_num, smaller_neighbours_num = integer()) {
    .Call('_dropestr_GetSmallerNeighboursDistributionsBySizes', PACKAGE = 'dropestr', dp_matrices, larger_neighbours_num, neighbour_prob_inds, size_adj, max_neighbour_num, smaller_neighbours_num)
}

GetSmallerNeighbourProbabilities <- function(small_neighs_dist, neighb_per_umi) {
    .Call('_dropestr_GetSmallerNeighbourProbabilities', PACKAGE = 'dropestr', small_neighs_dist, neighb_per_umi)
}

FilterUmisInGeneSimple <- function(reads_per_umi, neighbourhood, mult = 1) {
    .Call('_dropestr_FilterUmisInGeneSimple', PACKAGE = 'dropestr', reads_per_umi, neighbourhood, mult)
}

PredictNBC <- function(clf, predict_data) {
    .Call('_dropestr_PredictNBC', PACKAGE = 'dropestr', clf, predict_data)
}

PredictLeftPart <- function(neg_clf, rpu_prob, predict_data, gene_size) {
    .Call('_dropestr_PredictLeftPart', PACKAGE = 'dropestr', neg_clf, rpu_prob, predict_data, gene_size)
}

ArrangePredictions <- function(target_umi_factors, probs) {
    .Call('_dropestr_ArrangePredictions', PACKAGE = 'dropestr', target_umi_factors, probs)
}

FilterPredictions <- function(not_filtered_umis, base_umis, target_umis) {
    .Call('_dropestr_FilterPredictions', PACKAGE = 'dropestr', not_filtered_umis, base_umis, target_umis)
}

PrepareClassifierData <- function(reads_per_umi, neighborhood, umi_probabilities = numeric()) {
    .Call('_dropestr_PrepareClassifierData', PACKAGE = 'dropestr', reads_per_umi, neighborhood, umi_probabilities)
}

GetTrimCollisionsNum <- function(rpu_per_cell, trim_length) {
    .Call('_dropestr_GetTrimCollisionsNum', PACKAGE = 'dropestr', rpu_per_cell, trim_length)
}

GetBootstrapUmisMeanNum <- function(umi_probabilities, size, repeats_num, seed = -1L) {
    .Call('_dropestr_GetBootstrapUmisMeanNum', PACKAGE = 'dropestr', umi_probabilities, size, repeats_num, seed)
}

AdjustGeneExpressionClassic <- function(value, umis_number) {
    .Call('_dropestr_AdjustGeneExpressionClassic', PACKAGE = 'dropestr', value, umis_number)
}

#' Adjust gene expression value for collisions.
#'
#' @param value gene expression value.
#' @param observed_sizes vector of quantized ordered obseved gene sizes.
#' @param adjusted_sizes vector of adjusted gene sizes for *observed_sizes*.
#' @return Adjusted gene expression value.
#'
#' @export
AdjustGeneExpression <- function(value, observed_sizes, adjusted_sizes) {
    .Call('_dropestr_AdjustGeneExpression', PACKAGE = 'dropestr', value, observed_sizes, adjusted_sizes)
}

#' @export
ValueCountsC <- function(values) {
    .Call('_dropestr_ValueCountsC', PACKAGE = 'dropestr', values)
}

#' @export
ValueCounts <- function(values) {
    .Call('_dropestr_ValueCounts', PACKAGE = 'dropestr', values)
}

GetUmisDifference <- function(umi1, umi2, rpu1, rpu2, force_neighbours, umi_prob) {
    .Call('_dropestr_GetUmisDifference', PACKAGE = 'dropestr', umi1, umi2, rpu1, rpu2, force_neighbours, umi_prob)
}

BuildCountMatrix <- function(umis_per_gene) {
    .Call('_dropestr_BuildCountMatrix', PACKAGE = 'dropestr', umis_per_gene)
}

#' Parse UMIgs.
#'
#' @param reads_per_umigs data from the estimation step.
#' @param umi_length length of UMI.
#' @return List of lists of vectors with number of reads per UMI per gene per cell.
#'
#' @export
ParseUmisPerGene <- function(reads_per_umigs, umi_length) {
    .Call('_dropestr_ParseUmisPerGene', PACKAGE = 'dropestr', reads_per_umigs, umi_length)
}

TrimUmis <- function(rpu_per_cell, trim_length) {
    .Call('_dropestr_TrimUmis', PACKAGE = 'dropestr', rpu_per_cell, trim_length)
}

#' @export
AddIndexesToRpU <- function(reads_per_umi_per_cb, umis) {
    .Call('_dropestr_AddIndexesToRpU', PACKAGE = 'dropestr', reads_per_umi_per_cb, umis)
}

#' Estimate a distribution of observed UMI probabilities.
#'
#' @param umis_per_gene_per_cell list of vectors: number of UMIs per gene per cell (zeros can be omitted).
#' @param umi_length length of UMI.
#' @param smooth smooth term, which is added to each UMI probability in case if some UMIs have only few observations.
#' @return Vector of UMI probabilities.
#'
#' @export
GetUmisDistribution <- function(umis_per_gene_per_cell, smooth = 1L) {
    .Call('_dropestr_GetUmisDistribution', PACKAGE = 'dropestr', umis_per_gene_per_cell, smooth)
}

GetUmisList <- function(umi_len) {
    .Call('_dropestr_GetUmisList', PACKAGE = 'dropestr', umi_len)
}

ConcatLists <- function(lists) {
    .Call('_dropestr_ConcatLists', PACKAGE = 'dropestr', lists)
}

GetMirrorPairs <- function(pairs, probs, tol = 1e-5) {
    .Call('_dropestr_GetMirrorPairs', PACKAGE = 'dropestr', pairs, probs, tol)
}

