---
date: '`r Sys.Date()`'
output:
  html_document:
    df_print: kable
    theme: cerulean
  pdf_document: default
---

<!--
Input parameters:
  * d: data, obtained after dropEst run.
  * mit_genes: list of mitochondrion genes. Optional.
  * tags_data: list of data, obrained after dropTag run. Optional.
-->

<style>
  .float_left {
    float:left;
  }
  .float_right {
    float:right;
  }
</style>

```{r, include=FALSE}
knitr::opts_chunk$set(fig.width=5, fig.height=3, echo=FALSE, warning=FALSE, message=FALSE)
ggplot2::theme_set(ggplot2::theme_bw(base_size = 16) + ggplot2::theme(plot.title = ggplot2::element_text(hjust = 0.5)))
library(dropestr)
```

```{r}
umi_counts <- sort(Matrix::colSums(d$cm), decreasing=T)
```

## Common info
<div class="float_left">
```{r, fig.width=5.5, fig.height=3.5, include=FALSE, warning=FALSE}
PlotIntergenicFractionByChromosomes(d$reads_per_chr_per_cells, d$nonex_cells_chr_counts)
```
</div>
<div class="float_right">
```{r, fig.width=3.5, fig.height=3.5}
PlotUmisDistribution(d$reads_per_umi_per_cell)
```
</div>

<div class="float_left">
```{r, fig.width=4.5, fig.height=4.5}
par(mar = c(4, 4, 0.5, 0))
PlotReadsPerUmiByCells(d$mean_reads_per_umi, umi_counts, cex.lab=1.4)
```
</div>
<div class="float_right">
```{r, fig.width=4.5, fig.height=4.5}
PlotGenesPerCell(d$cm)
```
</div>

## Number of cells
<div class="float_left">
```{r, fig.width=4.5, fig.height=3.5}
PlotCellsNumberLine(umi_counts, breaks=80, title=NULL, estimate.cells.number=T)
```
</div>
<div class="float_right">
```{r, fig.width=4.5, fig.height=3.5}
PlotCellsNumberHist(umi_counts, breaks=60, estimate.cells.number=T, show.legend=F)
```
</div>
```{r, fig.width=5, fig.height=3}
PlotCellNumberLogLog(umi_counts, T, show.legend=F)
```

## Cells quality
```{r}
cells_quality <- EstimateCellsQuality(umi_counts)
bc_df <- PrepareLqCellsPipelineData(d, mitochondrion.genes = if (exists("mit_genes")) mit_genes else NULL, 
                                    total.reads.per.cb = if (exists("tags_data")) tags_data$reads_per_cb else NULL)
scores <- ScoreCells(bc_df, cells_quality)
```

<div class="float_left">
```{r, fig.width=4.5, fig.height=4}
PlotCellScores(scores, cells.num=EstimateCellsNumber(umi_counts), main='Cell scores')
```
</div>
<div class="float_right">
```{r, fig.width=4.5, fig.height=4}
if (exists("mit_genes")) {
  PlotMitochondrialFraction(d$cm, mit_genes, plot.threshold=T)
}
```
</div>

## Saturation
```{r, message=FALSE, warning=FALSE, fig.width=6, fig.height=4}
data(saturation_srr1784312)
saturation <- EstimateSaturation(d$reads_by_umig, d$reads_by_umig_cbs, sort(Matrix::colSums(d$cm), decreasing=T))
PlotSaturationEstimates(list(this=saturation, `mouse ES`=saturation_srr1784312))
```
